modules/core/include/opencv2/core/dualquaternion.inl.hpp

```diff
@@ -0,0 +1,392 @@
+        half_d * std::cos(half_angle) * qaxis;
+    return createFromQuat(Quat<T>::createFromAngleAxis(angle, Vec<T, 3>{qaxis[1], qaxis[2], qaxis[3]}), dual);
+}
+
+template <typename T>
+inline bool DualQuat<T>::operator==(const DualQuat<T> &q) const
+{
+    return (abs(w - q.w) < CV_DUAL_QUAT_EPS && abs(x - q.x) < CV_DUAL_QUAT_EPS &&
+            abs(y - q.y) < CV_DUAL_QUAT_EPS && abs(z - q.z) < CV_DUAL_QUAT_EPS &&
+            abs(w_ - q.w_) < CV_DUAL_QUAT_EPS && abs(x_ - q.x_) < CV_DUAL_QUAT_EPS &&
+            abs(y_ - q.y_) < CV_DUAL_QUAT_EPS && abs(z_ - q.z_) < CV_DUAL_QUAT_EPS);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getRealQuat() const
+{
+    return Quat<T>(w, x, y, z);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getDualQuat() const
+{
+    return Quat<T>(w_, x_, y_, z_);
+}
+
+template<typename T>
+inline DualQuat<T> DualQuat<T>::conjugate() const
+{
+    return DualQuat<T>(w, -x, -y, -z, w_, -x_, -y_, -z_);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::norm(QuatAssumeType assumeUnit) const
+{
+    Quat<T> real = getRealQuat();
+    T realNorm = real.norm();
+    if (assumeUnit)
+    {
+        return DualQuat<T>(realNorm, 0, 0, 0, 0, 0, 0, 0);
+    }
+    Quat<T> dual = getDualQuat();
+    if (realNorm < CV_DUAL_QUAT_EPS){
+        return DualQuat<T>(0, 0, 0, 0, 0, 0, 0, 0);
+    }
+    return DualQuat<T>(realNorm, 0, 0, 0, real.dot(dual) / realNorm, 0, 0, 0);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getRotation(QuatAssumeType assumeUnit) const
+{
+    if (assumeUnit)
+    {
+        return getRealQuat();
+    }
+    return getRealQuat().normalize();
+}
+
+template <typename T>
+Quat<T> DualQuat<T>::getTranslation(QuatAssumeType assumeUnit) const
+{
+    if (assumeUnit)
+    {
+        return 2.0 * getDualQuat() * getRealQuat().conjugate();
+    }
+    DualQuat<T> q = normalize();
+    return 2.0 * q.getDualQuat() * q.getRealQuat().conjugate();
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::normalize() const
+{
+    DualQuat<T> qNorm = norm();
+    if (qNorm.w < CV_DUAL_QUAT_EPS)
+    {
+        CV_Error(Error::StsBadArg, "Cannot normalize this dual quaternion: the norm is too small.");
+    }
+    return *this / qNorm;
+}
+
+template <typename T>
+inline T DualQuat<T>::dot(DualQuat<T> q) const
+{
+    return q.w * w + q.x * x + q.y * y + q.z * z + q.w_ * w_ + q.x_ * x_ + q.y_ * y_ + q.z_ * z_;
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::inv(QuatAssumeType assumeUnit) const
+{
+    if (assumeUnit){
+        return conjugate();
+    }
+    DualQuat<T> qNorm = norm();
+    if (qNorm.w < CV_DUAL_QUAT_EPS)
+    {
+        CV_Error(Error::StsBadArg, "This dual quaternion do not have inverse dual quaternion");
+    }
+    DualQuat<T> qNorm2 = qNorm * qNorm;
+    DualQuat<T> conj = conjugate();
+    Quat<T> p = conj.getRealQuat();
+    return createFromQuat(p / qNorm2.w, -p * qNorm2.w_ / (qNorm2.w * qNorm2.w ) + conj.getDualQuat() / qNorm2.w);
```

Comment:
Not sure if this formula is correct; if it's not, then this one can be used instead: `inv(a+e*b) = inv(a) - e*inv(a)*b*inv(a)`

Reply:
This formula is actually directly divided by its norm. 

The Taylor expansion is indeed more elegant. I'll replace it later